python/ray/experimental/client/common.py

```diff
@@ -17,83 +18,230 @@ def __repr__(self):
+        self.actor_cls = state["actor_cls"]
+        self._name = state["_name"]
+        self._ref = state["_ref"]
+
     def remote(self, *args, **kwargs):
         # Actually instantiate the actor
-        ref = ray.call_remote(self, ray_client_pb2.ClientTask.ACTOR, *args,
-                              **kwargs)
-        return ClientActorHandle(ref, self)
+        ref = ray.call_remote(self, *args, **kwargs)
+        return ClientActorHandle(ClientActorNameRef(ref.id), self)
 
     def __repr__(self):
-        return "ClientRemoteActor(%s, %s)" % (self._name, self.id)
+        return "ClientRemoteActor(%s, %s)" % (self._name, self._ref)
 
     def __getattr__(self, key):
+        if key not in self.__dict__:
+            raise AttributeError("Not a class attribute")
         raise NotImplementedError("static methods")
 
-
-class ClientActorHandle:
-    def __init__(self, actor_id: ClientActorRef,
+    def _prepare_client_task(self) -> ray_client_pb2.ClientTask:
+        if self._ref is None:
+            self._ref = ray.put(self.actor_cls)
+        task = ray_client_pb2.ClientTask()
+        task.type = ray_client_pb2.ClientTask.ACTOR
+        task.name = self._name
+        task.payload_id = self._ref.id
+        return task
+
+
+class ClientActorHandle(ClientStub):
+    """Client-side stub for instantiated actor.
+
+    A stub created on the Ray Client to represent a remote actor that
+    has been started on the cluster.  This class is allowed to be passed
+    around between remote functions.
+
+    Args:
+        actor_id: A reference to the running actor given to the client. This
+          is a serialized version of the actual handle as an opaque token.
+        actor_class: A reference to the ClientActorClass that this actor was
+          instantiated from.
+        _real_actor_handle: Cached copy of the Raylet-side ray.actor.ActorHandle
+          contained in the actor_id ref.
+    """
+
+    def __init__(self, actor_id: ClientActorNameRef,
                  actor_class: ClientActorClass):
         self.actor_id = actor_id
         self.actor_class = actor_class
+        self._real_actor_handle = None
+
+    def _get_ray_remote_impl(self):
+        if self._real_actor_handle is None:
+            self._real_actor_handle = ray.get_actor_from_object(self.actor_id)
+        return self._real_actor_handle
+
+    def __getstate__(self) -> Dict:
+        state = {
+            "actor_id": self.actor_id,
+            "actor_class": self.actor_class,
+            "_real_actor_handle": self._real_actor_handle,
+        }
+        return state
+
+    def __setstate__(self, state: Dict) -> None:
+        self.actor_id = state["actor_id"]
+        self.actor_class = state["actor_class"]
+        self._real_actor_handle = state["_real_actor_handle"]
 
     def __getattr__(self, key):
         return ClientRemoteMethod(self, key)
 
+    def __repr__(self):
+        return "ClientActorHandle(%s)" % (self.actor_class)
```

Comment:
Shouldn't it be the id instead of the class, to be similar to normal handle strs?

Reply:
Sure, it'll just be long and really the pickle though